reconcile/gitlab_members.py

```diff
@@ -0,0 +1,126 @@
+import sys
+import logging
+
+import utils.gql as gql
+from utils.config import get_config
+from utils.gitlab_api import GitLabApi
+
+GROUPS_QUERY = """
+{
+    groups: gitlabgroups_v1{
+        managedGroups
+    }
+}
+"""
+
+USERS_QUERY = """
+{
+    users: users_v1{
+        redhat_username
+        roles{
+            permissions{
+                ... on PermissionGitlabGroups_v1{
+                    name
+                    group
+                }
+            }
+        }
+    }
+}
+  """
+
+def get_gitlab_api():
+    config = get_config()
+    gitlab_config = config['gitlab']
+    server = gitlab_config['server']
+    token = gitlab_config['token']
+    return GitLabApi(server, token, ssl_verify=False)
+
+
+def get_groups():
+    gqlapi = gql.get_api()
+    groups = gqlapi.query(GROUPS_QUERY)['groups']
+    group_dict = {}
+    for g in groups[0]['managedGroups']:
+        group_dict[g] = []
+    return group_dict
+
+def get_desired_state():
+    gqlapi = gql.get_api()
+    gl = get_gitlab_api()
+    groups = gqlapi.query(GROUPS_QUERY)['groups']
+    users = gqlapi.query(USERS_QUERY)['users']
+    desired_group_members = get_groups()
+    for g in groups[0]['managedGroups']:
+        for u in users:
+            for r in u['roles']:
+                for p in r['permissions']:
+                    if 'group' in p and p['group'] == g:
+                        desired_group_members[g].append(gl.get_user(u['redhat_username'])) #does redhat_username == gitlab username? this should also account for if user was deleted
+        desired_group_members[g].append(gl.get_user('devtools-bot'))
+    return desired_group_members
+
+
+def get_current_state():
+    gqlapi = gql.get_api()
+    gl = get_gitlab_api()
+    groups = gqlapi.query(GROUPS_QUERY)['groups']
+    current_group_members = get_groups()
```

Comment:
this is an example how `get_groups` is not indicative of what it does (does it get the groups or the groups members?

Reply:
current_group_members is set to get_groups() which is a dict with all the groups as keys and empty lists as values, later the members are added to those empty lists.